File: SYP.Web/Modules/ServerTypes/Catalog/PriceListsForm.ts
import { BooleanEditor, DateEditor, EnumEditor, initFormType, LookupEditor, PrefixedContext, StringEditor, TextAreaEditor } from "@serenity-is/corelib";
import { PriceListItemsEditor } from "../../Catalog/PriceListItems/PriceListItemsEditor";
import { PriceListType } from "./PriceListType";

export interface PriceListsForm {
    Code: StringEditor;
    CurrencyId: LookupEditor;
    Name: StringEditor;
    Type: EnumEditor;
    Description: TextAreaEditor;
    ValidFrom: DateEditor;
    ValidTo: DateEditor;
    IsActive: BooleanEditor;
    IsDefault: BooleanEditor;
    ItemList: PriceListItemsEditor;
}

export class PriceListsForm extends PrefixedContext {
    static readonly formKey = 'Catalog.PriceLists';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!PriceListsForm.init) {
            PriceListsForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = EnumEditor;
            var w3 = TextAreaEditor;
            var w4 = DateEditor;
            var w5 = BooleanEditor;
            var w6 = PriceListItemsEditor;

            initFormType(PriceListsForm, [
                'Code', w0,
                'CurrencyId', w1,
                'Name', w0,
                'Type', w2,
                'Description', w3,
                'ValidFrom', w4,
                'ValidTo', w4,
                'IsActive', w5,
                'IsDefault', w5,
                'ItemList', w6
            ]);
        }
    }
}

[PriceListType]; // referenced types

File: SYP.Web/Modules/Catalog/PriceLists/PriceListsForm.cs
using Serenity.ComponentModel;

namespace SYP.Catalog.Forms;

[FormScript("Catalog.PriceLists")]
[BasedOnRow(typeof(PriceListsRow), CheckNames = true)]
public class PriceListsForm
{
    [Tab("General Information")]
    [HalfWidth, Placeholder("Leave blank to auto-generate")]
    public string Code { get; set; }

    [HalfWidth]
    public int CurrencyId { get; set; }

    [FullWidth]
    public string Name { get; set; }

    [FullWidth, TextAreaEditor(Rows = 3)]
    public string Description { get; set; }

    [HalfWidth, DateEditor]
    public DateTime ValidFrom { get; set; }

    [HalfWidth, DateEditor]
    public DateTime ValidTo { get; set; }

    [HalfWidth, DefaultValue(true)]
    public bool IsActive { get; set; }

    [HalfWidth]
    public bool IsDefault { get; set; }

    [Tab("Products")]
    [PriceListItemsEditor, SkipNameCheck]
    public List<PriceListItemsRow> ItemList { get; set; }
}